pyro/contrib/gp/models/svgp.py

```diff
@@ -0,0 +1,104 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.autograd import Variable
+from torch.distributions import constraints, transform_to
+from torch.nn import Parameter
+
+import pyro
+import pyro.distributions as dist
+from pyro.distributions.util import matrix_triangular_solve_compat
+
+from .vgp import VariationalGP
+
+
+class SparseVariationalGP(VariationalGP):
+    """
+    Sparse Variational Gaussian Process module.
+
+    References
+
+    [1] `Scalable variational Gaussian process classification`,
+    James Hensman, Alexander G. de G. Matthews, Zoubin Ghahramani
+
+    :param torch.autograd.Variable X: A 1D or 2D tensor of inputs.
+    :param torch.autograd.Variable y: A 1D tensor of outputs for training.
+    :param pyro.contrib.gp.kernels.Kernel kernel: A Pyro kernel object.
+    :param torch.Tensor Xu: An inducing-point parameter.
+    :param pyro.contrib.gp.likelihoods.Likelihood likelihood: A likelihood module.
+    :param float jitter: An additional jitter to help stablize Cholesky decomposition.
+    """
+    def __init__(self, X, y, kernel, Xu, likelihood, jitter=1e-6):
+        super(SparseVariationalGP, self).__init__(X, y, kernel, likelihood, jitter)
+
+        self.Xu = Parameter(Xu)
+        self.num_inducing = self.Xu.size(0)
+
+    def model(self):
+        self.set_mode("model")
+
+        kernel = self.kernel
+        likelihood = self.likelihood
+        Xu = self.get_param("Xu")
+
+        Kuu = kernel(Xu) + self.jitter.expand(self.num_inducing).diag()
+        Kuf = kernel(Xu, self.X)
+        Luu = Kuu.potrf(upper=False)
+
+        zero_loc = Variable(Xu.data.new([0])).expand(self.num_inducing)
+        # TODO: use scale_tril=Luu
+        u = pyro.sample("u", dist.MultivariateNormal(loc=zero_loc, covariance_matrix=Kuu))
+
+        # p(f|u) ~ N(f|mf, Kf)
+        # mf = Kfu @ inv(Kuu) @ u; Kf = Kff - Kfu @ inv(Kuu) @ Kuf
+        pack = torch.cat((u.unsqueeze(1), Kuf), dim=1)
+        Luuinv_pack = matrix_triangular_solve_compat(pack, Luu, upper=False)
+        Luuinv_u = Luuinv_pack[:, 0]
+        Luuinv_Kuf = Luuinv_pack[:, 1:]
+
+        mf = Luuinv_Kuf.t().matmul(Luuinv_u)
+        Kffdiag = kernel(self.X, diag=True)
+        Qffdiag = (Luuinv_Kuf ** 2).sum(dim=0)
+        Kfdiag = Kffdiag - Qffdiag
+
+        f = dist.Normal(mf, Kfdiag).sample()
```

Comment:
This is the place I mention about Monte Carlo with 1 sample. Did I make it right?